### PR TITLE
Add rollback support

### PR DESCRIPTION
**Pull request by syerrapragada**
@markandrus 
- Added rollback support for setLocalDescription

### CODE REVIEW

File: rtcpeerconnection.js [hunk 1]
```diff
@@ -865,10 +875,20 @@ module.exports = function(window, edgeVersion) {
       });
     }
 
-    pc._localDescription = {
-      type: description.type,
-      sdp: description.sdp
-    };
+    if (description.type === 'rollback') {
+      pc.transceivers.forEach(function(transceiver, sdpMLineIndex) {
+        transceiver.localCapabilitie = pc._previousLocalDescription;
```
Comment (markandrus) [inline rtcpeerconnection.js]: Typo?

**Comment by syerrapragada** [comment]
@manjeshbhargav @markandrus 
While testing edge implementation I found an issue in `setLocalDescription` where `_gather` is called even before the `_localDescription` is set resulting in `TypeError`. Last commit in this PR addresses this issue.

File: rtcpeerconnection.js [hunk 2]
```diff
@@ -802,73 +807,91 @@ module.exports = function(window, edgeVersion) {
...
+    if (description.type === 'rollback') {
+      pc._localDescription = pc._previousLocalDescription;
+      pc._previousLocalDescription = null;
 
       pc.transceivers.forEach(function(transceiver, sdpMLineIndex) {
-        pc._gather(transceiver.mid, sdpMLineIndex);
+        transceiver.localCapabilities = transceiver.previousLocalCapabilities;
+        transceiver.previousLocalCapabilities = null;
+        if (!pc._localDescription) {
+          transceiver.iceGatherer.onlocalcandidate = null;
```
Comment (markandrus) [inline rtcpeerconnection.js]: Why is this necessary?

File: test/rtcpeerconnection.js [hunk 3]
```diff
@@ -464,6 +483,42 @@ describe('Edge shim', () => {
         });
       });
     });
+
+    describe('called with rollback sdp type', () => {
+      let previousLocalDescription;
+      let previousLocalCapabilities;
+
+      beforeEach((done) => {
```
Comment (markandrus) [inline test/rtcpeerconnection.js]: Rather than thread this `done` through, you can just return the Promise chain started by `pc.createOffer`.

File: test/rtcpeerconnection.js [hunk 4]
```diff
@@ -464,6 +483,42 @@ describe('Edge shim', () => {
...
+          return pc.setLocalDescription({type: 'rollback'});
+        }).then(done);
+      });
+
+      it('should set the localDescription to previous description', () => {
+        expect(pc.localDescription).to.equal(previousLocalDescription);
+      });
+
+      it('should set localCapabilities to previous capabilities', () => {
+        let localCapabilities = pc.transceivers.map((transceiver) => {
```
Comment (markandrus) [inline test/rtcpeerconnection.js]: Prefer `const` to `let`.

File: rtcpeerconnection.js (hunk 2 above)
Comment (syerrapragada) [reply rtcpeerconnection.js]: `_gather` adds `onLocalCandidate` handler to iceGatherer objects. In `onLocalCandidate` callback when there is a candidate available, adds candidate to `_localDescription.sdp`.  In case of first offer rollback, `onLocalCandidate` throws error accessing sdp property of null description.

Comment (syerrapragada) [reply rtcpeerconnection.js]: I may need to restore https://github.com/twilio/rtcpeerconnection-shim/blob/6e4986ae27c542ce9e66f85657e2b351dddb20d3/rtcpeerconnection.js#L599 here

Comment (markandrus) [reply rtcpeerconnection.js]: If you're interacting with a proper EventTarget instance, then for any event "foo" emitted by some EventTarget instance, `instance`, the following event listener

```js
function listener() {}
instance.onfoo = listener;
instance.onfoo = null;
```

is independent of any event listeners added or removed via

```js
function listener() {}
instance.addEventListener('foo', listener);
instance.removeEventListener('foo', listener);
```

In the code you linked, I see a call to `addEventListener`. If your intention is to remove that, you should make the appropriate `removeEventListener` call, as setting `onlocalcandidate` to `null` is independent.
